datafusion/spark/src/function/string/format_string.rs

```diff
@@ -0,0 +1,2374 @@
+
+use arrow::{
+    array::{Array, ArrayRef, LargeStringArray, StringArray, StringViewArray},
+    datatypes::DataType,
+};
+use bigdecimal::{
+    num_bigint::{BigInt, Sign},
+    BigDecimal, ToPrimitive,
+};
+use chrono::{DateTime, Datelike, Timelike, Utc};
+use datafusion_common::{
+    exec_datafusion_err, exec_err, plan_err, DataFusionError, Result, ScalarValue,
+};
+use datafusion_expr::{
+    ColumnarValue, ScalarFunctionArgs, ScalarUDFImpl, Signature, TypeSignature,
+    Volatility,
+};
+
+/// Spark-compatible `format_string` expression
+/// <https://spark.apache.org/docs/latest/api/sql/index.html#format_string>
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct FormatStringFunc {
+    signature: Signature,
+    aliases: Vec<String>,
+}
+
+impl Default for FormatStringFunc {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FormatStringFunc {
+    pub fn new() -> Self {
+        Self {
+            signature: Signature::new(TypeSignature::VariadicAny, Volatility::Immutable),
+            aliases: vec![String::from("printf")],
+        }
+    }
+}
+
+impl ScalarUDFImpl for FormatStringFunc {
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn name(&self) -> &str {
+        "format_string"
+    }
+
+    fn aliases(&self) -> &[String] {
+        &self.aliases
+    }
+
+    fn signature(&self) -> &Signature {
+        &self.signature
+    }
+
+    fn return_type(&self, arg_types: &[DataType]) -> Result<DataType> {
+        if arg_types.is_empty() {
+            return plan_err!("The format_string function expects at least one argument");
+        }
+        if arg_types[0] == DataType::Null {
+            return Ok(DataType::Utf8);
+        }
+        if !matches!(
+            arg_types[0],
+            DataType::Utf8 | DataType::LargeUtf8 | DataType::Utf8View
+        ) {
+            return plan_err!("The format_string function expects the first argument to be Utf8, LargeUtf8 or Utf8View");
+        }
+        Ok(arg_types[0].clone())
+    }
+
+    fn invoke_with_args(&self, args: ScalarFunctionArgs) -> Result<ColumnarValue> {
+        let len = args
+            .args
+            .iter()
+            .fold(Option::<usize>::None, |acc, arg| match arg {
+                ColumnarValue::Scalar(_) => acc,
+                ColumnarValue::Array(a) => Some(a.len()),
+            });
+        let is_scalar = len.is_none();
+        let data_types = args.args[1..]
+            .iter()
+            .map(|arg| arg.data_type())
+            .collect::<Vec<_>>();
+        let fmt_type = args.args[0].data_type();
+
+        match &args.args[0] {
+            ColumnarValue::Scalar(ScalarValue::Null) => {
+                Ok(ColumnarValue::Scalar(ScalarValue::Utf8(None)))
+            }
+            ColumnarValue::Scalar(ScalarValue::Utf8(Some(fmt)))
+            | ColumnarValue::Scalar(ScalarValue::LargeUtf8(Some(fmt)))
+            | ColumnarValue::Scalar(ScalarValue::Utf8View(Some(fmt))) => {
```

Comment:
Need to handle case where its `ColumnarValue::Scalar(ScalarValue::Utf8(None))` (and for other string types too) I believe?